dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,475 @@
+            raise ValueError('n_jobs must be -1 (for full parallelization with '
+                             'dask) or 0 (for regular list comphrension). '
+                             'Setting n_jobs=0 can help memory overhead.')
+        self.n_jobs = n_jobs
+
+        est = model if not isinstance(model, ParallelPostFit) else model.estimator
+        if not hasattr(est, 'partial_fit'):
+            raise ValueError('Hyperband only supports models with partial_fit, '
+                             'which provides support for incremental learning. '
+                             'For non-incremental models, please consider '
+                             'RandomizedSearchCV instead')
+        if not hasattr(est, 'warm_start'):
+            warn('model has no attribute warm_start. Hyperband will assume it '
+                 'is reusing previous calls to `partial_fit` during each call '
+                 'to `partial_fit`')
+        else:
+            if not est.warm_start:
+                raise ValueError('Hyperband relies on model(..., warm_start=True).')
+
+        self.history = []
+
+        super_default = dict(iid=True, refit=True, cv=None,
+                             error_score='raise',
+                             return_train_score=_RETURN_TRAIN_SCORE_DEFAULT,
+                             scheduler=None, cache_cv=True)
+        for k, v in super_default.items():
+            if k in kwargs and kwargs[k] != super_default[k]:
+                msg = 'Hyperband ignores the keyword argument {k}={v}.'
+                warn(msg.format(k=k, v=v))
+        super(Hyperband, self).__init__(model, **kwargs)
+
+    def fit(self, X, y, dry_run=False, **fit_kwargs):
+        """
+        This function requires an active dask.distribtued client as it uses
+        the dask.distributed's concurrent.futures API.
+
+        This function implicitly assumes that higher scores are better. Note:
+        this matters if you are using a scoring function that measures loss,
+        where lower values imply a better model.
+
+        This function works with classes with a certain API, not only
+        scikit-learn estimators. It is assumed that the class
+
+        * implements ``partial_fit(self, X, y, **kwargs)``.
+        * implements ``set_params(self, **kwargs) -> self``
+        * implements ``score(self, X, y) -> number``
+        * supports ``sklearn.base.clone(self, safe=True)``. Support for this
+          function comes down to having an function ``get_params(self)``.
+
+        """
+        self.scorer_, _ = self._check_scorer(self.model)
+
+        client = _get_client()
+        if client:
+            import distributed
+            variables = {'val_score': -np.inf, 'config': None}
+
+            shared = {}
+            for key, value in variables.items():
+                # distributed.Variable requires client
+                shared[key] = distributed.Variable(key)
+                shared[key].set(value)
```

Comment:
It would be useful to remove the dependence on dask.distributed if easy.

Reply:
I tried to remove it and preserve the global state but couldn't. The only benefits to the global state is notifying the user of the best validation loss and the associated parameters.

This benefit is marginal. I've removed dask.distributed, and print+log the best validation loss and associated parameters.

There's an extension of Hyperband that does require the distributed scheduler, and has larger benefits too (interrupting model selection can still find the best model). I'll push any benefits to that.